feat(language-service): Quick fix to import a component when its selector is used

dylhunn opened the pull request:
The language service can now generate an import corresponding to a selector. This includes both the TypeScript module import and the decorator import. This applies to both standalone components and components declared in NgModules.

Previous PRs in this sequence:
Language Service: https://github.com/angular/angular/pull/47181
Compiler: https://github.com/angular/angular/pull/47180, https://github.com/angular/angular/pull/47166, https://github.com/angular/angular/pull/47561, https://github.com/angular/angular/pull/47631

atscott commented on packages/language-service/src/codefixes/utils.ts:
```diff
@@ -114,6 +115,45 @@ export function convertFileTextChangeInTcb(
+/**
+ * Determine whether this node is a TS Object Literal Expression node.
+ */
+export function isObjectNode(node: tss.Node): node is tss.ObjectLiteralExpression {
+  return node.kind === ts.SyntaxKind.ObjectLiteralExpression;
+}
```
ts.isObjectLiteralExpression does this. There are usually helper functions for all the `node.kind === xyz` of the form `isXYZ`

atscott commented on packages/language-service/src/codefixes/utils.ts:
```diff
@@ -114,6 +115,45 @@ export function convertFileTextChangeInTcb(
+/**
+ * Determine whether the provided object literal node has a particular key.
+ */
+export function objectNodeHasKey(node: tss.ObjectLiteralExpression, key: string): boolean {
```
nit: s/objectNode/objectLiteral

atscott commented on packages/language-service/src/codefixes/fix_missing_import.ts:
```diff
@@ -0,0 +1,115 @@
+ */
+function insertDecoratorArgument(
+    component: tss.ClassDeclaration, importCmp: string): tss.FileTextChanges[] {
+  const fileTextChanges: tss.FileTextChanges[] = [];
```
I would actually remove this const. It's unclear in all your early returns that you're actually returning a set of no changes. They should just return `[]` instead and your last line here can return the one change.

atscott commented on packages/language-service/src/codefixes/fix_missing_import.ts:
```diff
@@ -0,0 +1,115 @@
+    return fileTextChanges;
+  }
+
+  const insertImportChange = addPropertyToObjectNode(
+      objArgs, component.getSourceFile().fileName, importsKey, `[${importCmp}]`);
+  fileTextChanges.push(insertImportChange);
+
+  return fileTextChanges;
```
```suggestion
  return [addPropertyToObjectNode(
      objArgs, component.getSourceFile().fileName, importsKey, `[${importCmp}]`)];
```

atscott commented on packages/language-service/src/codefixes/fix_missing_import.ts:
```diff
@@ -0,0 +1,115 @@
+    const missingComponentClass = 'ThreeCmp';
+    const missingComponentPath = 'one/two/three.ts';
+
+    // TODO(dylhunn): Is there an elegant way to generate this (e.g. serializing a synthetic AST?)
```
Yes. Check out https://ts-ast-viewer.com/ (thanks @pkozlowski-opensource)

https://ts-ast-viewer.com/#code/JYWwDg9gTgLgBAbziYBnVwB2BzAwhcCTAU0xlwBsBDdOAXzgDMoC4ByFdLPAyEsgApUYACzYBuIA

The bottom gives you the ts factory methods you'll need to call.

atscott commented on packages/language-service/src/codefixes/fix_missing_import.ts:
```diff
@@ -0,0 +1,115 @@
+    const importString = `import { ${missingComponentClass} } from '${missingComponentPath}';\n`;
+    const fileImportChange = {
+      fileName: templateInfo.component.getSourceFile().fileName,
+      // TODO(dylhunn): Just insert at the top; can we do something fancier with the AST?
```
This is probably fine. You could consider inserting after the last import statement, before the first, etc. `findFirstMatchingNode` can help here.

atscott commented on packages/language-service/src/codefixes/utils.ts:
```diff
@@ -114,6 +115,45 @@ export function convertFileTextChangeInTcb(
+ * Determine whether the provided object literal node has a particular key.
+ */
+export function objectNodeHasKey(node: tss.ObjectLiteralExpression, key: string): boolean {
+  const argSyntaxList = node.getChildren().filter(c => c.kind === ts.SyntaxKind.SyntaxList);
+  if (argSyntaxList.length !== 1) return false;
+  const argPropertyAssignments =
+      argSyntaxList[0].getChildren().filter(c => c.kind === ts.SyntaxKind.PropertyAssignment);
+  return argPropertyAssignments.flatMap(a => a.getChildren()).some(c => c.getText().includes(key));
```
You should use astexplorer.net to help with some of these things.

https://astexplorer.net/#/gist/6ba115801f80f1d406ced01602c370c2/150b10d1fd8f8bc8a3d7d9b6192a86b6e83d0fbf

`node.properties` here will give you the property assignments.

And then you'll probably want to use `.name` instead of `getChildren` to get the name.

I've written something just like this recently:

```
node.properties.find(
        (p): p is ts.PropertyAssignment =>
            isPropertyAssignmentWithName(p, 'children'));

function isPropertyAssignmentWithName(
    property: ts.ObjectLiteralElementLike,
    name: string): property is ts.PropertyAssignment {
  return ts.isPropertyAssignment(property) && ts.isIdentifier(property.name) &&
      property.name.text === name;
}
```

dylhunn replied:
This will be fixed in a separate PR: https://github.com/angular/angular/pull/47181

dylhunn committed:
feat(language-service): Quick fix to import a component when its selector is used

The language service can now generate an import corresponding to a selector. This includes both the TypeScript module import and the decorator import. This applies to both standalone components and components declared in NgModules.

eneajaho commented:
Hi @dylhunn, great job 🥳
From my understanding and the test cases in this PR, does it mean this will work only for importing components and not directives and pipes? Or will work also for importing all of them?

dylhunn commented:
@eneajaho The current implementation on the language service side only affects quick fixes for selectors in component positions. However, all the plumbing I have added will be reused in v15.1 for as-you-type autocomplete, and selectors in other positions (e.g. attributes).

atscott commented on packages/language-service/src/codefixes/fix_missing_import.ts:
```diff
@@ -0,0 +1,169 @@
+          potentialImport.moduleSpecifier}' on ${importOn.name?.text}`,
+      changes: [componentImportChange, fileImportChange],
+    }];
+    return codeActions.map(codeAction => {
+      return {
+        fixName: codeAction.fixName,
+        fixId: codeAction.fixId,
+        fixAllDescription: codeAction.fixAllDescription,
+        description: codeAction.description,
+        changes: convertFileTextChangeInTcb(codeAction.changes, compiler),
+        commands: codeAction.commands,
+      };
+    });
```
This `map` is a bit weird to me. How about just doing `[componentImportChange, fileImportChange].map(...)` above instead of mapping the whole object to change a single property?

dylhunn replied:
It turns out we don't need this `map` at all! I We're not actually making any changes in the typecheck block at all.

atscott commented on packages/language-service/src/codefixes/fix_missing_import.ts:
```diff
@@ -0,0 +1,169 @@
+    });
+  },
+  fixIds: [FixIdForCodeFixesAll.FIX_MISSING_IMPORT],
+  // 'Fix All' is not supported, because the imports may come from different files and should be
+  // examined individually.
```
Maybe I'm not understanding something here, but don't you still individually analyze each diagnostic in the `getAllCodeActions` function?

dylhunn replied:
Right -- I'm not saying the results would be incorrect, just that each quick fix could potentially import a different file. It seemed harmful to allow bulk-importing a large number of different suggested files without the user verifying them. But I'm overthinking it and we should just support this as well.

atscott replied:
For Typescript's fix missing imports, it gives you the option to select the correct one if you want to fix a single one. If you select the "fix all", then it just does the first one.

![image](https://user-images.githubusercontent.com/479713/194167757-968f77ed-e5c9-4ded-965f-57f6641d483a.png)

Selecting the "add all missing imports" item from the above actions results in 
```
import { PostModule } from 'post';

import {ABCD, AppComponent} from './app.component';
```

ivanwonder replied:
I agree with @atscott 

BTW, I think the right way to disable `fix all` is to remove the `fixId` from `CodeAction`.

https://github.com/microsoft/TypeScript/blob/58bae8db69b275a3efa57b14b486778c55185552/src/services/types.ts#L790-L794

dylhunn replied:
Thanks @ivanwonder. On second thought I agree as well it should be fine to provide `fix all`.

atscott commented on packages/language-service/src/ts_utils.ts:
```diff
@@ -5,6 +5,7 @@
  * Use of this source code is governed by an MIT-style license that can be
  * found in the LICENSE file at https://angular.io/license
  */
+import {findAllMatchingNodes} from '@angular/compiler-cli/src/ngtsc/typecheck/src/comments';
```
nit: I think this import isn't right. I don't think we do deep imports to `src` from the `@angular/...` namespace.

dylhunn replied:
So this one isn't needed, but I have the same import in `fix_missing_import.ts`. What is the right fix for this? `findAllMatchingNodes` doesn't actually appear to be re-exported anywhere in the compiler. Do you think it would be appropriate to add a re-export somewhere, or should I not be using this at all?

atscott replied:
@alxhub I'm not sure what we should do here

atscott commented on packages/language-service/src/ts_utils.ts:
```diff
@@ -129,3 +130,124 @@ export function updateObjectValueForKey(
+ * omitted.
+ */
+export function generateImport(
+    rawName: string, rawPropertyName: string|null,
+    rawModuleSpecifier: string): ts.ImportDeclaration {
```
suggestion: Based on the documentation above, my suggested renames for the arguments are:
rawName => aliasImportName (and make this nullable)
rawPropertyName => exportedSpecifierName (and make this not nullable)
rawModuleSpecifier => moduleName

I think this also means the logic below needs to be slightly adjusted

Same goes for the `updateImport` function below

dylhunn replied:
So I actually chose these specific names because they are the same as the names used in the TypeScript AST. Still happy to change them if you prefer -- I just wanted to point this out explicitly, since I think there's some value in having the matching names.

atscott replied:
I do think it's super weird that `rawPropertyName` is the exported identifier from the module when `rawName` is supplied but `rawName` is the exported identifier when `rawPropertyName` is omitted.

But I think this is TypeScript's fault. If you made the change I suggested, the code below would be all wonky to get it working with the ts API.

what about:
rawName => localName
rawPropertyName => exportedSpecifierName

Up to you. I understand now why it's like this.

dylhunn replied:
I've adopted these new names, but the semantics are still aligned with TS.

dylhunn commented on packages/language-service/src/codefixes/fix_missing_import.ts:
```diff
@@ -0,0 +1,169 @@
+
+    // The error must be an invalid element in tag, which is interpreted as an intended selector.
+    const target = getTargetAtPosition(templateInfo.template, start);
+    if (target?.context.kind !== TargetNodeKind.ElementInTagContext) return [];
```
I will add the braces everywhere in this PR.

dylhunn commented on packages/language-service/src/codefixes/fix_missing_import.ts:
```diff
@@ -0,0 +1,169 @@
+    const target = getTargetAtPosition(templateInfo.template, start);
+    if (target?.context.kind !== TargetNodeKind.ElementInTagContext) return [];
+    const missingComponentSelector = (target.context.node as any).name as string | undefined;
+    if (!missingComponentSelector) return [];
```
Fixed.

dylhunn commented on packages/language-service/src/codefixes/fix_missing_import.ts:
```diff
@@ -0,0 +1,169 @@
+    };
+    if (existingImportName === null) {
+      const existingImportDecl = allImports.find(
+          decl => decl.moduleSpecifier.getText().includes(potentialImport.moduleSpecifier));
```
The reason strict equality doesn't work is because the raw module specifier will be `'./foo'`, but the suggested specifier will be `./foo`. Notice that the former has quotes, and the latter does not. I have fixed this in the spirit of your suggestion by specifically removing quote characters, rather than doing a general substring comparison.

dylhunn commented on packages/language-service/src/codefixes/fix_missing_import.ts:
```diff
@@ -0,0 +1,169 @@
+    }
+
+    // Update the imports in the decorator's `imports` array.
+    const componentImportChange =
```
Indeed!

dylhunn commented on packages/language-service/src/codefixes/fix_missing_import.ts:
```diff
@@ -0,0 +1,169 @@
+    if (potentialImports.length === 0) return [];
+    const potentialImport = potentialImports[0];
+
+    // The trait might already be imported, possibly under a different name. If so, determine the
+    // local name of the imported trait.
+    const allImports =
+        findAllMatchingNodes(importOn.getSourceFile(), {filter: ts.isImportDeclaration});
+    const existingImportName: string|null =
+        importsHave(allImports, potentialImport.symbolName, potentialImport.moduleSpecifier);
```
Yes, that was a bug. That's what I get for sending a PR review before I've written all my unit tests. 🤦

The existing `hasImport` helper from g3 doesn't appear to handle namespace imports, which we want. So I'm leaving this thread open for now, and I'll try to modify it.

dylhunn commented on packages/language-service/src/codefixes/fix_missing_import.ts:
```diff
@@ -0,0 +1,169 @@
+    // In the future, we could handle multiple matches as additional quick fixes.
+    const allPossibleDirectives = checker.getPotentialTemplateDirectives(templateInfo.component);
+    const matchingDirectives =
+        allPossibleDirectives.filter((d) => d.selector === missingComponentSelector);
```
I had not given much thought to the presence of attribute selectors yet, but I'm glad this already exists, and have added a generic for `T extends {selector: string}` (since that's the only real constraint on the argument shape).

dylhunn replied:
(And indeed, the quick fix now appears even for complex selectors like `<standalone-boo[attr] some-other-directive>`!)

atscott commented on packages/language-service/src/ts_utils.ts:
```diff
@@ -129,3 +129,164 @@ export function updateObjectValueForKey(
+
+export function moduleSpecifierHasPath(moduleSpecifier: ts.Expression, path: string): boolean {
+  return moduleSpecifier
+             .getText()
+             // Remove the wrappping quotes from the module specifier text.
+             .replace(/["']/g, '') === path;
```
again taking inspiration from the internal `hasImport` implementation, I think this would be better as:

```
ts.isStringLiteral(moduleSpecifier) &&
        moduleSpecifier.text === importPath
```

dylhunn replied:
Ahh yeah, this is better.

atscott commented on packages/language-service/src/codefixes/fix_missing_import.ts:
```diff
@@ -0,0 +1,165 @@
+    if (matchingDirectives.size === 0) {
+      return [];
+    }
+    const bestMatch = [...matchingDirectives][0];
```
```suggestion
    const bestMatch = matchingDirectives[0];
```
Not sure if there was a reason for the array copy here? I'm guessing there was a refactor at some point

dylhunn replied:
Indeed.

atscott commented on packages/language-service/src/codefixes/fix_missing_import.ts:
```diff
@@ -0,0 +1,165 @@
+      const importName = nonCollidingImportName(allImports, potentialImport.symbolName);
+
+      if (existingImportDeclaration !== undefined) {
+        // Update an existing import declaration.
+        const bindings = existingImportDeclaration.importClause?.namedBindings;
+        if (bindings === undefined || ts.isNamespaceImport(bindings)) {
+          // This should be impossible, as a namespace imports are already existing for any symbol.
+          return [];
+        }
+        let span = {start: bindings.getStart(), length: bindings.getWidth()};
+        const updatedBindings = updateNamedImport(bindings, potentialImport.symbolName, importName);
+        const importString = printNode(updatedBindings, importOn.getSourceFile());
+        fileImportChange.textChanges.push({span, newText: importString});
+      } else {
+        // Generate a new import declaration, and insert it at the top of the file.
+        const newImport =
+            generateImport(potentialImport.symbolName, importName, potentialImport.moduleSpecifier);
+        const importString = printNode(newImport, importOn.getSourceFile()) + '\n';
+        fileImportChange.textChanges.push({span: {start: 0, length: 0}, newText: importString});
+      }
```
```suggestion
    fileImportChange.textChanges.push(someHelper(...))
```
Would it make sense to extract this chunk to a helper function that returns the additional text change?

atscott commented on packages/language-service/src/codefixes/fix_missing_import.ts:
```diff
@@ -0,0 +1,165 @@
+ */
+export const missingImportMeta: CodeActionMeta = {
+  errorCodes,
+  getCodeActions: (
+      {templateInfo, start, compiler, formatOptions, preferences, errorCode, tsLs}) => {
```
Can we actually extract the function here so this looks more like:

```
export const missingImportMeta: CodeActionMeta = {
  errorCodes,
  getCodeActions,
  fixIds: [FixIdForCodeFixesAll.FIX_MISSING_IMPORT],
  getAllCodeActions,
};
```

And move it to the very top or bottom of the file?

alxhub commented on packages/language-service/src/codefixes/fix_missing_import.ts:
```diff
@@ -0,0 +1,179 @@
+
+  // The error must be an invalid element in tag, which is interpreted as an intended selector.
+  const target = getTargetAtPosition(templateInfo.template, start);
+  if (target === null || target?.context.kind !== TargetNodeKind.ElementInTagContext ||
```
Nit: safe navigation should not be necessary here (and in general, prefer not using it for conditionals as it can make it extremely tricky to reason about what's actually being checked).

dylhunn replied:
Removed, and ack, I'll dial it back in the future.

alxhub commented on packages/language-service/src/codefixes/fix_missing_import.ts:
```diff
@@ -0,0 +1,179 @@
+  const missingElement = target.context.node;
+
+  // The class which has an imports array; either a standalone trait or its owning NgModule.
+  const importOn = checker.getOwningNgModule(templateInfo.component) ?? templateInfo.component;
```
Nit: what if this is a component which isn't `standalone: true` but isn't in an NgModule?

We should prefer to check specifically that it's marked `standalone`.

alxhub commented on packages/language-service/src/codefixes/fix_missing_import.ts:
```diff
@@ -0,0 +1,179 @@
+  const codeActions: ts.CodeFixAction[] = [{
+    fixName: FixIdForCodeFixesAll.FIX_MISSING_IMPORT,
+    description:
+        `Import ${importName} from '${potentialImport.moduleSpecifier}' on ${importOn.name?.text}`,
```
Using `?.` here is sketchy as it could show the string `undefined` to the user. Prefer `!.` - at least it'll crash.

alxhub commented on packages/language-service/src/codefixes/fix_missing_import.ts:
```diff
@@ -0,0 +1,179 @@
+
+  // The trait might already be imported, possibly under a different name. If so, determine the
+  // local name of the imported trait.
+  const allImports = findAllMatchingNodes(file, {filter: ts.isImportDeclaration});
```
This feels like the wrong operation - `findAllMatchingNodes` is a utility designed to scan a TCB, not a `ts.SourceFile`. Consider writing your own utility here for finding an existing import of a module specifier specifically.

Note that there can be multiple ways to import files. E.g. the same file may be imported as `@app/path/to/file` or as `../relative/path/to/file`, so we can't always rely on string comparison here. A much better option is to use the `ts.TypeChecker` to `getSymbolAtLocation` and actually check if the import points to the file we want it to. Even this isn't really foolproof because of re-exports.

dylhunn replied:
Fixed as suggested.

alxhub commented on packages/language-service/src/codefixes/fix_missing_import.ts:
```diff
@@ -0,0 +1,179 @@
+    }
+    let span = {start: bindings.getStart(), length: bindings.getWidth()};
+    const updatedBindings = updateImport(bindings, newImport.symbolName, importName);
+    const importString = printNode(updatedBindings, file);
```
Using the `ts.Printer` is probably okay for now, but we should invest a bit in this in the future. Ideally we would figure out the correct textual change without printing out the node, which might destroy its existing formatting.

dylhunn replied:
Ack. We certainly can generate text insertions based on existing node positions, without reserializing the whole thing, but then we're doing string manipulation instead of using `ts.factory`. So tradeoffs.

alxhub commented on packages/language-service/src/codefixes/fix_missing_import.ts:
```diff
@@ -0,0 +1,179 @@
+  const newImportDeclaration =
+      generateImport(newImport.symbolName, importName, newImport.moduleSpecifier);
+  const importString = printNode(newImportDeclaration, file) + '\n';
+  return [[{span: {start: 0, length: 0}, newText: importString}], importName];
```
Are we putting the import at the start of the file here? We shouldn't do that, as there may be license comments, etc. above the code.

Instead, we should look for the last existing import, and insert the new one after it.

alxhub commented on packages/language-service/src/ts_utils.ts:
```diff
@@ -129,3 +129,165 @@ export function updateObjectValueForKey(
+ */
+export function ensureArrayWithIdentifier(
+    identifier: ts.Identifier, arr?: ts.ArrayLiteralExpression): ts.ArrayLiteralExpression|null {
+  if (!arr) {
```
Nit: prefer an explicit `!== undefined` - don't rely on JS truthiness coercion.

alxhub commented on packages/language-service/src/ts_utils.ts:
```diff
@@ -129,3 +129,165 @@ export function updateObjectValueForKey(
+  if (!arr) {
+    return ts.factory.createArrayLiteralExpression([identifier]);
+  }
+  if (arr.elements.find(v => ts.isIdentifier(v) && v.text === identifier.text)) {
```
Nit: same.

alxhub commented on packages/language-service/src/ts_utils.ts:
```diff
@@ -129,3 +129,165 @@ export function updateObjectValueForKey(
+ */
+function importHas(importDecl: ts.ImportDeclaration, propName: string): string|null {
+  const bindings = importDecl.importClause?.namedBindings;
+  if (bindings === undefined) return null;
```
Nit: always use braces:
```suggestion
  if (bindings === undefined) {
    return null;
  }
```

alxhub commented on packages/language-service/src/ts_utils.ts:
```diff
@@ -129,3 +129,165 @@ export function updateObjectValueForKey(
+             .filter(
+                 declaration =>
+                     moduleSpecifierHasPath(declaration.moduleSpecifier, moduleSpecifier))
+             .map(decl => importHas(decl, propName))
```
Nit: be consistent about names of variables here (`declaration` vs `decl`) and avoid single letter names like `p`.

alxhub commented on packages/language-service/src/ts_utils.ts:
```diff
@@ -129,3 +129,165 @@ export function updateObjectValueForKey(
+  // First, we handle the case of explicit named imports.
+  if (ts.isNamedImports(bindings)) {
+    // The name of the property in its export scope might differ from its local name.
+    const nameInExportScope = (i: ts.ImportSpecifier) => i.propertyName?.text ?? i.name.text;
```
Use a utility function with a declared type, not a `const` here.

alxhub commented on packages/language-service/src/ts_utils.ts:
```diff
@@ -129,3 +129,165 @@ export function updateObjectValueForKey(
+    // name.
+    const specifier = bindings.elements.find(i => propName == nameInExportScope(i));
+    // Return the name of the property in the *local* scope.
+    return specifier?.name.text ?? null;
```
Avoid safe navigation in cases like this.

alxhub commented on packages/language-service/src/ts_utils.ts:
```diff
@@ -129,3 +129,165 @@ export function updateObjectValueForKey(
+ * Generator function that yields an infinite sequence of alternative aliases for a given symbol
+ * name, starting with the name itself.
+ */
+function* suggestAlternativeSymbolNames(name: string): Iterator<string> {
```
👏

alxhub commented on packages/language-service/src/ts_utils.ts:
```diff
@@ -129,3 +129,165 @@ export function updateObjectValueForKey(
+ * name, starting with the name itself.
+ */
+function* suggestAlternativeSymbolNames(name: string): Iterator<string> {
+  yield name;
```
Do we need to yield the original `name` here as it seems like `nonCollidingImportName` checks it already, before using the generator.

dylhunn replied:
True :)

alxhub commented on packages/language-service/test/code_fixes_spec.ts:
```diff
@@ -226,6 +226,53 @@ describe('code fixes', () => {
+      'foo.ts': ` import {CommonModule} from '@angular/common';
+      import {Component} from '@angular/core';
+
+      @Component({selector: 'foo', template: '<bar></bar>', standalone: true})
```
I'm concerned that our inserter will butcher the formatting of existing `@Component` structures. How well does it work if every property is on its own indented line, like 99.9% of existing Angular code?

dylhunn replied:
Hand testing shows that the line breaks don't get messed up (at least, not in any case I have tried). I will add a test that checks the whitespace explicitly, to ensure the formatting is OK in the common case.

Formatting in the language service in general needs more attention, possibly as part of my work extending this project in Q4.

dylhunn commented:
@alxhub I have applied fixes for all your comments, thanks!

jessicajaniuk commented:
This PR was merged into the repository by commit bebef5fb43dcf54a109499d0d9dd701786c33f60.
